atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/CqlClientFactory.java

```diff
@@ -0,0 +1,159 @@
+
+        sslOptions(config).ifPresent(clusterBuilder::withSSL);
+
+        Cluster cluster = clusterBuilder.build();
+
+        ThreadFactory threadFactory = new ThreadFactoryBuilder()
+                .setNameFormat(cluster.getClusterName() + "-session" + "-%d")
+                .build();
+
+        return CqlClientImpl.create(cluster, Executors.newCachedThreadPool(threadFactory), initializeAsync);
+    }
+
+    private static Optional<RemoteEndpointAwareJdkSSLOptions> sslOptions(CassandraKeyValueServiceConfig config) {
+        if (config.sslConfiguration().isPresent()) {
+            SSLContext sslContext = SslSocketFactories.createSslContext(config.sslConfiguration().get());
+            return Optional.of(RemoteEndpointAwareJdkSSLOptions.builder()
+                    .withSSLContext(sslContext)
+                    .build());
+        } else if (config.ssl().isPresent() && config.ssl().get()) {
+            return Optional.of(RemoteEndpointAwareJdkSSLOptions.builder().build());
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    private static PoolingOptions poolingOptions(CassandraKeyValueServiceConfig config) {
+        return new PoolingOptions()
+                .setMaxConnectionsPerHost(HostDistance.LOCAL, config.poolSize())
+                .setMaxConnectionsPerHost(HostDistance.REMOTE, config.poolSize())
+                .setPoolTimeoutMillis(config.cqlPoolTimeoutMillis());
+    }
+
+    private static QueryOptions queryOptions(CassandraKeyValueServiceConfig config) {
+        return new QueryOptions().setFetchSize(config.fetchBatchCount());
+    }
+
+    private static LoadBalancingPolicy loadBalancingPolicy(CassandraKeyValueServiceConfig config,
+            Set<InetSocketAddress> servers) {
+        // Refuse to talk to nodes twice as (latency-wise) slow as the best one, over a timescale of 100ms,
+        // and every 10s try to re-evaluate ignored nodes performance by giving them queries again.
+        // Note we are being purposely datacenter-irreverent here, instead relying on latency alone
+        // to approximate what DCAwareRR would do;
+        // this is because DCs for Atlas are always quite latency-close and should be used this way,
+        // not as if we have some cross-country backup DC.
+        LoadBalancingPolicy policy = LatencyAwarePolicy.builder(new RoundRobinPolicy()).build();
+
+        // If user wants, do not automatically add in new nodes to pool (useful during DC migrations / rebuilds)
+        if (!config.autoRefreshNodes()) {
+            policy = new WhiteListPolicy(policy, servers);
+        }
+
+        // also try and select coordinators who own the data we're talking about to avoid an extra hop,
+        // but also shuffle which replica we talk to for a load balancing that comes at the expense
+        // of less effective caching
+        return new TokenAwarePolicy(policy, TokenAwarePolicy.ReplicaOrdering.RANDOM);
```

Comment:
This is pretty interesting, thanks for explaining your reasoning. 
- Not the scope of this, but could be interesting to see if later on we can have an AZ aware policy. In a sense, load balancing should be taken care of by http-remoting clients talking to service nodes.
- @tpetracca / @amandachow if you guys have a bit of time, your input would be valuable here; this sounds reasonable, but not sure if it's right.

Reply:
nit: TokenAwarePolicy.ReplicaOrdering.RANDOM is the default.